Design/WorkspacePropertyWithIndex.md

```diff
@@ -0,0 +1,137 @@
+	- `IndexType::WorkspaceIndex` for list of global workspace indices typically from 0-N.
+	- `IndexType::DetectorID` for list of detector ids within the instrument.
+2. Users will be restricted to entering only one of the allowed index types at the GUI level. 
+3. Validation will occur in `IndexInfo::makeIndexSet`(which throws for bad indices), not in every algorithm.
+4. The `ArrayProperty` will be exposed e.g `WorkspacePropertyWithIndex::getArrayProperty` in order to make use of `ListPropertyWidget`.
+5. Overload function call operator to return `std::pair`.
+6. `IndexInfo` will be used to translate the indices at the point of access i.e. `WorkspacePropertyWithIndex::operator()()`.
+7. There must be a few modifications to the GUI generation code in order to support this type:
+	-  `PropertyWidgetFactory` may need to handle `WorkspacePropertyWithIndex` specifically.
+	-  May need to introduce a concept of a group of property widgets.
+
+<sup>1</sup><sub>The distinction is made here between local (a subset of all data partitioned on an MPI Rank) and global (entire data set). This will not be visible at the GUI or client API level.</sub>
+
+### Example usage
+
+**Property Declaration**:
+
+Declaring the property in C++:
+```cpp
+//Property name may be fixed?
+declareProperty(WorkspacePropertyWithIndex("InputWorkspace", IndexType::SpectrumNumber|IndexType::WorkspaceIndex));
+//IndexType::WorkspaceIndex is the defaul
+declareProperty(WorkspacePropertyWithIndex("InputWorkspace"));
+```  
+In order to reduce typing we could include builder functions which allow for more concise property declaration:
+```cpp
+//Builder method returns a unique_ptr?
+declareProperty(CreateMatrixWorkspaceWithSpectrumNumber("InputWorkspaceWithIndex");
+//or alternatively
+declareProperty(CreateWithSpectrumNumber<MatrixWorkspace>("InputWorkspaceWithIndex");
+```
+
+**Algorithm Dialog Box (GUI)**:
+
+**InputWorkspace** [ _________________ ]<br>
+( **.** ) **SpectrumNumber** (  ) **WorkspaceIndex**<br>
+**Indices** [ _________________ ]
+ 
+**Client Code Access**:
+
+*C++*:
+
+Setting the property manually (tuple is created internally):
+
+```cpp
+setProperty("InputWorkspaceWithIndex", ws, std::vector<SpectrumNumber>{1, 2, 3, 4});
+//or
+setProperty("InputWorkspaceWithIndex", ws, IndexType::SpectrumNumber, "1:33,42");
+```
+
+Accessing the property:
+
+```cpp
+MatrixWorkspace_const_sptr inputWS;
+IndexSet indices;
+std::tie(inputWS, indices) = getProperty("InputWorkspaceWithIndex");//returns tuple
+```
+
+*Python*:
+
+
+In order to reduce the load on script authors, we could split keyword arguments and create types which capture each index type. This expanding arguments allows users to understand the autocomplete in the scripting editor more easily.:
+```python
+# The index types could accept a range as a string or
+# a numpy array which contains the values.
+ChangePulseTime(TimeOffset=10, InputWorkspace=ws, Spectra=("1:33") )
```

Comment:
I agree that I think this is the best option from a user perspective as it introduces the fewest new concepts :+1: